### PR TITLE
Support UTF sequences in stream gobbler

### PR DESCRIPTION
**Pull request by algairim**
Signed-off-by: Mykola Mandra <mykola.mandra@cloudsoft.io>

### CODE REVIEW

File: utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java [hunk 1]
```diff
@@ -60,79 +63,121 @@ public void shutdown() {
...
+
     public StreamGobbler setLogPrefix(String prefix) {
         logPrefix = prefix;
         return this;
-    }    
-    
+    }
+
     @Override
     public void run() {
```
Comment (algairim) [inline utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java:84]: Key changes are in `run()` method, everything else is formatting.

File: utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java [hunk 2]
```diff
@@ -60,79 +63,121 @@ public void shutdown() {
...
-        int c = -1;
+        int c, bytes = 0;
+        char[] utfSymbol = new char[2];
         try {
-            while (running.get() && (c=stream.read())>=0) {
```
Comment (algairim) [inline utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java:81]: To revert `running.get() &&`!

Comment (algairim) [reply utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java:81]: ab0c367f6e76e87f5e876678d033a9c2f31195ac

File: utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java [hunk 3]
```diff
@@ -60,50 +63,84 @@ public void shutdown() {
...
+
+                //System.out.println(" [" + bb.position() + "] " + Integer.toHexString(c)); // for DEBUG
+                bb.put((byte) c);
+                bytes--;
+
+                if (bytes == 0) {
+                    bb.rewind();
+                    StandardCharsets.UTF_8.decode(bb).get(utfSymbol);
+                    bb.clear();
+                    //System.out.println("[ch] 0x" + Integer.toHexString(utfSymbol[0]) + " 0x" + Integer.toHexString(utfSymbol[1])); // for DEBUG
```
Comment (iuliana) [inline utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java:112]: @algairim  If this is useful put it in a trace log statement.

File: utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java [hunk 4]
```diff
@@ -60,50 +63,84 @@ public void shutdown() {
...
+                    } else if (c >= 0xE0) {
+                        bytes = 3;
+                    } else if (c >= 0xC2) {
+                        bytes = 2;
+                    } else {
+                        bytes = 1;
+                    }
+                }
+
+                //System.out.println(" [" + bb.position() + "] " + Integer.toHexString(c)); // for DEBUG
```
Comment (iuliana) [inline utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java:104]: @algairim  If this is useful put it in a trace log statement.

File: utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java [hunk 5]
```diff
@@ -23,17 +23,20 @@
...
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 import org.slf4j.Logger;
 
 public class StreamGobbler extends Thread implements Closeable {
-    
+
+    private static char REPLACEMENT_CHARACTER = 0xfffd;
```
Comment (iuliana) [inline utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java:34]: @algairim Make this final.

File: utils/common/src/test/java/org/apache/brooklyn/util/stream/StreamGobblerTest.java [hunk 6]
```diff
@@ -29,12 +29,65 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.Assert;
 import org.testng.annotations.Test;
 
 public class StreamGobblerTest {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamGobblerTest.class);
+
     private String NL = Os.LINE_SEPARATOR;
```
Comment (iuliana) [inline utils/common/src/test/java/org/apache/brooklyn/util/stream/StreamGobblerTest.java:43]: @algairim Make this final.

**Review by iuliana — CHANGES_REQUESTED** [review]
A few comments related to logging and static variabled that should be final. Otherwise looks good.

File: utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java (hunk 5 above)
Comment (algairim) [reply utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java]: Addressed in ba780b1acac6c42b4d4f5cc8aa0d063339869847.

File: utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java (hunk 4 above)
Comment (algairim) [reply utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java]: This code is not needed anymore, removed in ba780b1acac6c42b4d4f5cc8aa0d063339869847.

File: utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java (hunk 3 above)
Comment (algairim) [reply utils/common/src/main/java/org/apache/brooklyn/util/stream/StreamGobbler.java]: This code is not needed anymore, removed in ba780b1acac6c42b4d4f5cc8aa0d063339869847.

File: utils/common/src/test/java/org/apache/brooklyn/util/stream/StreamGobblerTest.java (hunk 6 above)
Comment (algairim) [reply utils/common/src/test/java/org/apache/brooklyn/util/stream/StreamGobblerTest.java]: Addressed in ba780b1acac6c42b4d4f5cc8aa0d063339869847.

**Comment by iuliana** [comment]
Failures unrelated to this commit. Will merge.

**Pull request merged**